extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1328,53 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
-    ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (this.isPaywallStory_()) {
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          return this.blockOnPendingSubscriptionsState_(
+            targetPageId,
+            direction
+          );
+        }
+
+        // Attempt to navigate to the locked pages after the paywall page.
+        if (pageIndex > PAYWALL_PAGE_INDEX) {
+          this.pageAfterGranted_ = targetPageId;
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+          // Show the paywall immediately if the active page is the paywall page.
+          if (this.isOnPaywallPage_()) {
+            this.showSubscriptionsDialog_();
+            return Promise.resolve();
+          }
+
+          // Redirect to the paywall page.
+          return this.switchTo_(
+            this.pages_[PAYWALL_PAGE_INDEX].element.id,
+            direction
+          );
+        }
+
+        if (this.pageAfterGranted_) {
+          this.showSubscriptionsDialog_();
```

Comment:
I have no idea how you can get there, does this mean `if (SubscriptionState.BLOCKED && && pageIndex > PAYWALL_PAGE_INDEX)`?
The block setting `pageAfterGranted_` above has an early return, so this code never runs?

Reply:
The early return above would redirect/switch to the paywall page since we only show the paywall on paywall page, e.g. switchTo(page-3) calls switchTo(paywall-page). Inside switchTo(paywall-page) at this line, the ageAfterGranted is not null so show the paywall and don't return early to finish the remaining switchTo operations to paywall page.

Reply:
Made obsolete by separating out the logic of showing the UI and switch to the paywall page.